locale.c

```diff
@@ -6294,129 +6255,307 @@ S_my_langinfo_i(pTHX_
-                                     NULL, &this_is_utf8);
-                Safefree(scratch_buf);
-                if (this_is_utf8 == UTF8NESS_NO) {
-                    is_utf8 = UTF8NESS_NO;
-                    break;
-                }
+        /* We can also try various strings associated with LC_TIME, like the
+         * names of months or days of the week */
 
-                if (this_is_utf8 == UTF8NESS_YES) {
-                    is_utf8 = UTF8NESS_YES;
-                }
+            DAY_1, DAY_2, DAY_3, DAY_4, DAY_5, DAY_6, DAY_7,
+            MON_1, MON_2, MON_3, MON_4, MON_5, MON_6, MON_7, MON_8,
+                                        MON_9, MON_10, MON_11, MON_12,
+            ALT_DIGITS, AM_STR, PM_STR,
+            ABDAY_1, ABDAY_2, ABDAY_3, ABDAY_4, ABDAY_5, ABDAY_6, ABDAY_7,
+            ABMON_1, ABMON_2, ABMON_3, ABMON_4, ABMON_5, ABMON_6,
+            ABMON_7, ABMON_8, ABMON_9, ABMON_10, ABMON_11, ABMON_12
+
+#          endif
+        };
+
+#          ifdef USE_LOCALE_TIME
+
+        /* The code in the recursive call below can handle switching the
+         * locales, but by doing it now here, that code will check and discover
+         * that there is no need to switch then restore, avoiding those each
+         * loop iteration */
+        const char * orig_TIME_locale = toggle_locale_c(LC_TIME, locale);
+
+#          endif
+
+        /* The trials array may consist of strings from two different locale
+         * categories.  The call to my_langinfo_i() below needs to pass the
+         * proper category for each string.  There is a max of 1 trial for
+         * LC_MONETARY; the rest are LC_TIME.  So the array is arranged so the
+         * LC_MONETARY item (if any) is first, and all subsequent iterations
+         * will use LC_TIME.  These #ifdefs set up the values for all possible
+         * combinations. */
+#          if defined(USE_LOCALE_MONETARY) && defined(HAS_LOCALECONV)
+
+        locale_category_index  cat_index = LC_MONETARY_INDEX_;
+
+#            ifdef USE_LOCALE_TIME
+
+        const locale_category_index  follow_on_cat_index = LC_TIME_INDEX_;
+        assert(trials[1] == DAY_1); /* Make sure only a single non-time entry */
+
+#            else
+
+        /* Effectively out-of-bounds, as there is only the monetary entry */
+        const locale_category_index  follow_on_cat_index = LC_ALL_INDEX_;
+
+#            endif
+#          elif defined(USE_LOCALE_TIME)
+
+        locale_category_index  cat_index = LC_TIME_INDEX_;
+        const locale_category_index  follow_on_cat_index = LC_TIME_INDEX_;
+
+#          else
+
+        /* Effectively out-of-bounds, as here there are no trial entries at
+         * all.  This allows this code to compile, but there are no strings to
+         * test, and so the answer will always be non-UTF-8. */
+        const locale_category_index  cat_index =
+                                     follow_on_cat_index = LC_ALL_INDEX_;
```

Comment:
What declares `follow_on_cat_index` in this case?

Also, `cat_index` is const here, the loop below may never actually execute but it still tries to modify cat_index.

I suspect this should have been:
```
        locale_category_index  cat_index = LC_ALL_INDEX_;
        const locale_category_index  follow_on_cat_index = LC_ALL_INDEX_;
```

Reply:
Yes, good catch.  I hadn't bothered to actually build this configuration; but I did with your fixes, and it passed its tests